gax/src/main/java/com/google/api/gax/rpc/mtls/MtlsProvider.java

```
@@ -0,0 +1,191 @@
+      // file doesn't exist
+      return null;
+    }
+  }
+
+  @VisibleForTesting
+  static KeyStore getKeyStore(InputStream metadata, ProcessProvider processProvider)
+      throws IOException {
+    try {
+      Process process = processProvider.createProcess(metadata);
+
+      // Run the command and timeout after 1000 milliseconds.
+      int exitCode = runCertificateProviderCommand(process, 1000);
+      if (exitCode != 0) {
+        throw new IOException("Cert provider command failed with exit code: " + exitCode);
+      }
+
+      // Create mTLS key store with the input certificates from shell command.
+      return SecurityUtils.createMtlsKeyStore(process.getInputStream());
+    } catch (InterruptedException e) {
+      throw new IOException("Interrupted executing certificate provider command", e);
+    } catch (GeneralSecurityException e) {
+      throw new IOException("Failed to get key store", e);
+    }
+  }
+
+  @VisibleForTesting
+  static ImmutableList<String> extractCertificateProviderCommand(InputStream contextAwareMetadata)
+      throws IOException {
+    JsonParser parser = new GsonFactory().createJsonParser(contextAwareMetadata);
+    ContextAwareMetadataJson json = parser.parse(ContextAwareMetadataJson.class);
+    return json.getCommands();
+  }
+
+  @VisibleForTesting
+  static int runCertificateProviderCommand(Process commandProcess, long timeoutMilliseconds)
+      throws IOException, InterruptedException {
+    long startTime = System.currentTimeMillis();
+    long remainTime = timeoutMilliseconds;
+
+    while (remainTime > 0) {
+      Thread.sleep(Math.min(remainTime + 1, 100));
+      remainTime -= System.currentTimeMillis() - startTime;
+
+      try {
+        // Check if process is terminated by polling the exitValue, which throws
+        // IllegalThreadStateException if not terminated.
+        return commandProcess.exitValue();
+      } catch (IllegalThreadStateException ignored) {
```

Curious, why is it safe to ignore this? Do we need to tell the caller about this?

---

Good. add this information to the comment

---

done

---

Move the comment into the catch block so as not to trip up static analysis tools that warn about empty catch blocks.

---

done